### DISCUSSION TITLE
[Proposal] Ignored Directives Support

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#8616] [Language Ideas]
# Ignored Directives

* [x] Proposed: https://github.com/dotnet/csharplang/blob/main/proposals/ignored-directives.md
* [ ] Prototype
* [ ] Implementation
* [ ] Specification

## Summary
[summary]: #summary

As we move to allow C# statements in the top level of a file, it's reasonable to assume that we might someday get support for `dotnet run`ing a .cs file. For further use of C# as a scripting language in this vein, I propose that we introduce a new preprocessor directive who's purpose is to exist solely to be ignored by the language: the hashbang directive. Linux and Unix scripts will often use `#!<path to interpreter>` in order to direct the program loader to the interpreter that can run them. We should add a similar ignore to the C# language.

## Detailed design
[design]: #detailed-design

We introduce a new directive:

```antlr
pp_directive
    : ... // Existing directives
    | pp_hashbang
    | pp_reference
    ;

pp_hashbang
    : whitespace? '#' '!' input_character*
    ;

pp_reference
    : whitespace? '#' 'r' whitespace input_character*
    ;
```

The language ignores the contents of the hashbang and reference directives and moves to the next line.

## Drawbacks
[drawbacks]: #drawbacks

Any change adds complication to the language. If we implement support for this and it starts appearing in files without actual scripting support, then it could be confusing for users expecting it to work.

## Alternatives
[alternatives]: #alternatives

N/A

## Unresolved questions
[unresolved]: #unresolved-questions

1. If we use these directives to support tooling such as nuget references, we will need to consider how they interact with other pragmas, such as `#if`. We have a few options for this:
    1. Require that ignored directives are the first directives in a file. We could even go further and require they're all at the beginning of the file.
    2. Less restrictive than 1, but forbid ignored directives inside conditional directives.
    3. Least restrictive, require a true preprocessing step in the compiler. Today, we can process directives up front without having to do a full multi-step preprocessing pass before lexing, but if we want to support nuget directives inside conditional directives we may need the full two-step process.

## Design meetings

- https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-05-12.md#simple-c-programs
- https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-08-31.md#ignored-directives-support
- https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-09-26.md#ungrouped

**Comment by ufcpp** [comment]
I'm worried about BOM. [The presence of the BOM before the shebang will prevent the script interpreter](https://en.wikipedia.org/wiki/Shebang_(Unix)).

Visual Studio for Win uses Encoding.Default (usually not UTF8) if the BOM doesn't present.

**Comment by 333fred** [comment]
Sure, and we could certainly consider warning if the file uses anything other than UTF-8 w/o BOM.

**Comment by yaakov-h** [comment]
I thought shebang was already supported... is it only in the .csx variant?

**Comment by yaakov-h** [comment]
ah, looks like its only currently supported in the scripting mode.

https://github.com/dotnet/roslyn/blob/1b266a1e67434578ec09737a4e9427e52f44a420/src/Compilers/CSharp/Portable/Parser/DirectiveParser.cs#L108-L111

**Comment by svick** [comment]
1. As @yaakov-h mentions, shebang is already supported in the scripting dialect. Is this effectively formalizing that support and making it work in both dialects?
2. If I understand it correctly, the shebang that's supported by the scripting mode has to be literally the first thing in the file: no whitespace before it, no other lines before it. Since that's the only way shebang works, does it make sense to support whitespace before it and to support it on lines other than the first line in the file?

**Comment by HaloFour** [comment]
I'm a little confused.  Do "simple programs" not require a project file?  How would a single file "script" in this manner establish references or change any compiler properties?

This feels like trying to bridge a gap between C# "simple programs" and CSX, which already has this support plus the directive support to make it actually useful.  Short of adding that directive support this feels extremely limited.

Again, C# would benefit so much more from embracing CSX and improving the tooling support than it ever will by chasing CSX through some separate dialect that will always be much less capable.

**Comment by dersia** [comment]
I agree with @HaloFour without directives for assembly imports this might not be very useful. On the other hand having a script that downloads nuget packages or loading complex and somewhat big assemblies doesn't feel right for scripting.

The only thing that I kind of see as a big plus is that you can just take the cs file to any machine architecture and call `dotnet run` on it so you don't have to compile aot.

What is still unclear for me is if the purpose of this proposal is that the cs-file is being interpreted or is it going to be compiled on running it. 

If the purpose is to "just" interpreted on run, than I agree even more with @HaloFour we should focus on making csx a better experience.

**Comment by hez2010** [comment]
I don't think it should introduce a new `#!` directive to C#, for C# doesn't use `#` for comment syntax.

How about:
```csharp
/// <interpreter path="" />
```
Which will allow more complex definition, for example:
```csharp
/// <project target="net5.0">
///   <properties>
///     <nullable>enable</nullable>
///   </properties>
///   <items>
///     <package include="Newtonsoft.Json" version="12.0.3" />
///   </items>
/// </project>
```

**Comment by HaloFour** [comment]
The point is to use a syntax that existing shells already recognize,
otherwise there would be no benefit.  Unix shells recognize shebang syntax
and it is widely used to enable a script file to instruct the shell in how
that file is to be interpreted.  No shell will understand a C# comment or a
blob of XML.

**Comment by aka-nse** [comment]
I summarize my understanding of this proposal at followings. Is it right?

- C# compiler will simply ignore a shebang in the first line of the source. Because the lines start with `#` are preprocessor directives even on conventional spec, this modification looks not to impact other features.

- You can write inline csproj XML on and after 2nd line, but the interpretation of them should be done by build tool (e.g. dotnet-cli), not compiler. Such an inline csproj will be written under the conventional comment syntax, and the extension spec will be defined by tools.

```csharp
#!/bin/dotnet run
//+ <Project Sdk="Microsoft.NET.Sdk">
//+     <PropertyGroup>
//+         <TargetFramework>netcoreapp3.1</TargetFramework>
//+     </PropertyGroup>
//+ </Project>

/*
 * In this example, it is assumed that dotnet-cli will interprets the comments start with `//+` as inline csproj
 */

using System;

Console.WriteLine("Hello, world!");
```

**Comment by 333fred** [comment]
The first point is correct, and is the only thing this proposal covers. This proposal makes no statement on how a program would communicate dependencies to a tool.

**Comment by rolfbjarne** [comment]
One of the design meetings says:

> Would be nice to have, but only in conjunction with the rest of the simple programs feature. Implementation might be one of the simplest ever features in the compiler, but without the tooling to accompany it we don't know that this is the right design.

I'm not sure I agree, because this doesn't actually need official tooling to be useful, it's entirely possible to extend by anyone.

Consider this:

```cs
#!/bin/my-custom-program

using System;

Console.WriteLine("Hello, world!");
```

where `/bin/my-custom-program` can be a program that takes the script and does whatever it wants to - which could be creating a project file for the file and running `dotnet run`, but it an also be anything else.

One example could be:


```cs
#!/bin/run-unit-test

[TestFixture]
public class Test {
    [Test]
    public void IsValid (int number)
    {
        Assert.AreEqual (42, number, "everything");
    }
}
```

where `run-unit-test` would create a project, add the corresponding NUnit references and usings, compile & run the tests.

**Comment by CyrusNajmabadi** [comment]
> it's entirely possible to extend by anyone.

The point is that without that tooling we don't know if this will be good enough to actually solve those cases.

**Comment by rolfbjarne** [comment]
> > it's entirely possible to extend by anyone.
> 
> The point is that without that tooling we don't know if this will be good enough to actually solve those cases.

I agree that the reference directive (`#r`) might be in this category, but the shebang directive (`#!`) can stand on its own and would be useful by itself, both without the reference directive, or any other tooling.

**Comment by CyrusNajmabadi** [comment]
> > > it's entirely possible to extend by anyone.
> > 
> > The point is that without that tooling we don't know if this will be good enough to actually solve those cases.
> 
> I agree that the reference directive (`#r`) might be in this category, but the shebang directive (`#!`) can stand on its own and would be useful by itself, both without the reference directive, or any other tooling.

What is it's not good enough?  What if we do it, and a month later people try to use it and say it's not sufficient and it needs to change or be redone. 

We can't add features without some way of actually validating that it solves the problems we want it to solve. We need some actual tooling adoption to demonstrate it is sufficient.

**Reply by colejohnson66** [reply]
This argument is illogical. A shebang is a well-known directive in Linux. _How_ could it not be "good enough"? Your argument is circular. You say something might not be "good enough" but that you'd need "actual tooling adoption". This isn't C; We don't have other compilers implementing "extensions" that the language body can graft from.

It's quite annoying how proposals are often "not good enough", but when some other Microsoft sub-org [wants to do something](https://github.com/dotnet/sdk/pull/46915), the language is [updated/mandated to support it](https://github.com/dotnet/csharplang/pull/9192).

**Reply by HaloFour** [reply]
I agree, this is a slippery slope.  The C# language can't do anything with that directive.  It will never understand or respect the interpreter, nor any of the additional directives used by said interpreter.  IMO, all of that should be the responsibility of the tooling.  The interpreter should strip that directive (and all others) before feeding the C# source to the compiler.  Any IDE supporting this interpreter should do the same.  If someone stuck said source in the middle of a project not meant to be used by an interpreter, that should fail, not silently succeed because the compiler is ignoring it.

**Reply by CyrusNajmabadi** [reply]
> It's quite annoying how proposals are often "not good enough", but when some other Microsoft sub-org https://github.com/dotnet/sdk/pull/46915, the language is https://github.com/dotnet/csharplang/pull/9192.

I'm nto sure how that's annoying.  In this case we *would* be able to see if it sufficient.  This is exactly what i was referring to in my post.  We don't want to add features without a way of actually validating them.  The linked proposal allows us a path to actually validate the feature.  And, as such, meets my criteria for inclusion.

**Reply by colejohnson66** [reply]
Because Microsoft has the capital to invest in writing code for a feature that doesn't exist yet. And then they have the willpower to force C# to adapt. We don't.

How does `dotnet run file.cs` validate a feature any more than the other examples in this thread?

@HaloFour does make a good argument that hasn't been brought up. One downside, however, is that many tools expect actual files. So, writing a `dotnet run` or `/bin/run-unit-tests` without ignored directives would require trimming them out, then serializing the output to disk before invoking the real tool. A minor overhead that ignored directives would be able to remove.

**Reply by HaloFour** [reply]
From the conversation on Discord it sounds like none of the teams want to have to deal with virtual files or rewriting.  They'd rather the compiler support and ignore these directives enough to support this use case without affecting tooling.  I can understand that.

What I don't get is the desire to move C# in this direction.  CSX already exists, with relatively little uptake to my knowledge.  It's also much better suited for this role as it already supports directives to reference outside packages and doesn't have the weird limitations that come with top-level statements.  If there is a desire to support C# as a scripting language (or as a single file self-contained project) I'd rather see that established with some clear goals.

**Reply by CyrusNajmabadi** [reply]
>How does dotnet run file.cs validate a feature any more than the other examples in this thread?

Because we actually have a scenario we can validate.  Like we can literally try out `dotnet run ...` and see if the solution proposed is sufficient for its needs, or if it needs work.

**Reply by yaakov-h** [reply]
Why could the team not do this with `dotnet-script` i.e. `#!/usr/bin/env dotnet script` or whatever the actual syntax is? Or with a *nix build of `csi.exe`?

**Reply by jkrejcha** [reply]
I guess my fundamental question here is, in what way is `#pragma` not good enough for this use case ("validating the use case of a feature")? More useful and fundamental things to the language are specified this way (`#pragma warning disable/restore`).

In addition, what use is `#sdk`, `#package` to an implementation that doesn't use these things (most of .NET when using project-based projects (>99% of them), stuff compiled with `csc`, some alternative compiler, etc) such that it has to be a core language feature, especially for what's best as I can tell, an experiment?

---

If `#pragma` isn't good enough, there is also the alternative of just supporting "conditionally-supported-directive" as in C and C++ as well like so

> A sequence of preprocessing tokens is only a conditionally-supported-directive if it does not begin with any of the directive names appearing after a # in the syntax.
> A conditionally-supported-directive is conditionally-supported with implementation-defined semantics.

C# isn't currently like this as far as I can tell ([although .NET effectively treats it this way](https://github.com/dotnet/csharplang/discussions/8616#discussioncomment-12395065)) but this would work too.

**Reply by jaredpar** [reply]
> What I don't get is the desire to move C# in this direction. CSX already exists, with relatively little uptake to my knowledge.

Part of the problem with CSX is that it is not C#: it is a different dialect of the language with different semantics. That made it challenging to port to C# as you had to reconcile these differences. Once CSX projects hit the point where scripting wasn't enough and they needed a full fledged program they were at a bit of an impass. There wasn't a simple "change this to a C# console project" switch that they could leverage. 

One of the explicit goals of `dotnet run app.cs` is that there is a _clean_ and simple grow up story to a real project file app. The `dotnet project add` command will convert to a real project with no loss of fidelity. Many of the questions that we have in the design come back to "how does this work with `dotnet project add`"?

**Reply by jaredpar** [reply]
>  They'd rather the compiler support and ignore these directives enough to support this use case without affecting tooling. I can understand that.

The desire is that these be supported in the parser and the compiler host decide what to do, or not to do with them.

**Reply by HaloFour** [reply]
> Part of the problem with CSX is that it is not C#: it is a different dialect of the language with different semantics. That made it challenging to port to C# as you had to reconcile these differences. Once CSX projects hit the point where scripting wasn't enough and they needed a full fledged program they were at a bit of an impass.

Indeed, and those differences in the dialect are what make CSX so much better at being a scripting language.  In C# you have to deal with those awkward rules of what and where you can declare anything.  You're sorta implicitly in a method body, and you're declaring local functions, but you can also declare types, but only in certain places.  IMO, it's an awful experience, and short of continuing to try to reconcile those dialectal differences it's likely to remain one.

I also think it's a niche case of a niche case.  What percentage of people really want to use C# as a pseudo-scripting language, and then what percentage of those people have a desire to flip to a full-fledged project after the fact?

**Reply by colejohnson66** [reply]
As a bit of a tangent, one of the most common uses for single-file projects is probably @dotnet/runtime itself. Almost every test (in [tests](https://github.com/dotnet/runtime/tree/main/src/tests)) is a two-file system: the program or test itself, and a(n almost) copy-pasted `csproj` file.

**Reply by jaredpar** [reply]
> I also think it's a niche case of a niche case. 

I disagree. Seen plenty of cases over the years where functionality, particularly in infra, starts off as a simple script. Over time it evolves and gain features to the point that scripting can no longer really handle it and you need a real program. The migration can be quite time consuming. 

> Indeed, and those differences in the dialect are what make CSX so much better at being a scripting language.

At the time CSX was created a I agree with this assertion. It was **much** easier to write simple programs in CSX vs. C#. Now though with top level statements + local functions the gap's narrowed significantly.

**Reply by HaloFour** [reply]
> At the time CSX was created a I agree with this assertion. It was much easier to write simple programs in CSX vs. C#. Now though with top level statements + local functions the gap's narrowed significantly.

Sure, it's simpler, but it's still awkward due to how top-level statements are implemented and the opinions they enforce, such as requiring that all statements must precede any type declarations.  CSX is so much more flexible.

I wish there would be a much more holistic and intentional initiative to support C# in these scenarios rather than bolting things onto the side.

**Reply by jjonescz** [reply]
> I wish there would be a much more holistic and intentional initiative to support C# in these scenarios rather than bolting things onto the side.

I think top level statements, dotnet run file, etc are all part of one initiative to support writing simple C# programs.

> such as requiring that all statements must precede any type declarations.

That might very well change in a future C# version as part of this initiative.

**Reply by jkrejcha** [reply]
To be honest, I think it's probably more likely that this doesn't happen as much (things like PowerShell which you can call into the .NET runtime or whatever, Python, or even just something like sh or $shell_of_choice) are extremely popular and they tend to fill a different niche than C# proper. CSX has found its place, but it is pretty niche. The usefulness of that can be debated, but I think it's still missing a couple of very important things:

I guess my problem is still the point of **"this feature already exists"** and **"this makes no sense for >99% of C# programs and would be barely be used in the most popular implementation of C# let alone any others"**. 

There's nothing fundamentally stopping .NET's implementation from recognizing `pragma sdk` today and doing something/nothing with it. They don't need support from the language because they control their own implementation thereof, all the compiler has to do is... just not do anything when it encounters any directives other tooling the implementation cares about knows what to do with.

The entire point of `pragma` (and conditionally-supported directives in C and C++) is to do things in an implementation-defined way, which imo makes for a much better way to experiment than just throwing random stuff into the spec (and if it gets used it does and if not (the default), it sits there having to be maintained forever). There's really nothing gained by standardizing `#sdk` and `#package`, because the language doesn't know about this (and shouldn't, imo).

**Reply by jjonescz** [reply]
> There's really nothing gained by standardizing `#sdk` and `#package`, because the language doesn't know about this (and shouldn't, imo).

Wouldn't the same apply to `#region`? I'd also argue that conciseness of `#package` vs `#pragma package` is an important reason. Especially considering this feature is intended for writing *simple* C# programs.

**Reply by HaloFour** [reply]
You could make that argument, and honestly if someone had suggested `#region` much more recently it'd probably be a tougher sell.  But at least it does have a specific meaning and use case, plus a known syntax (e.g. `#endregion` required).

The language can't know what these other directives mean, or if they have a specific structure, unless the language becomes much more bound to this specific tooling.  It can't even know what directives are possible, or might be possible in a future release.  Supporting a pattern for those directives and treating them like comments is probably the safest bet.  But at the same time is there a reason that the tooling _must_ use new directives like this?

**Reply by jaredpar** [reply]
> There's nothing fundamentally stopping .NET's implementation from recognizing pragma sdk today and doing something/nothing with it.

There is nothing stopping us from recognizing `// package ...` or a half dozen other syntaxes. Language design isn't just about finding what combinations of existing syntax we can use but also deciding will that combination make sense for users? I would argue that customers are fairly ingrained into believing that `#pragma` is a way to control diagnostics in C#. Extending that to control other scenarios is likely to be less intuitive vs. a new syntax specifically geared to it.

**Reply by jkrejcha** [reply]
Given the language (rightly) doesn't define the exact characteristics of diagnostics other than that certain compiler warnings/errors may, must, or mustn't occur, warning control seems reasonable to be in that implementation-defined bucket. `#diagnostic restore <implementation-defined>` is no better than `#pragma warning restore FOO001`. In fact, I'd argue it's in fact worse since with the pragma, it's easier to see "this is dependent on what .NET does" without having to read the spec (as most programmers don't).

The problem I see is that `#sdk` and `#package` doesn't really have a purpose except to be an implementation-defined comment (that most implementations don't really care about), which already exists in the language. I'm not necessarily opposed to conditionally-supported directives as a general feature of the language, but these 2 feels too specific to a single **potential** implementation

**Reply by jkrejcha** [reply]
Side note here: the compiler does recognize magic attributes like `DllImportAttribute` and some tooling does recognize both attributes (`StringSyntaxAttribute`) and magic comments.

The prior art here, even in the .NET ecosystem, usually goes for attributes (or in some cases pragmas or comments). The attribute method is actually nicer in some respects since it potentially adds reflectable metadata, but it also has an advantage of not being tied to the implementation (the language doesn't have to know about shared libraries, the host operating system, etc).

I'm not saying this is the best idea for CSX, but it's better than the proposed method imo

**Reply by HaloFour** [reply]
If we did want to keep it to shorter preprocessor directives I'd rather the language not adopt these specified directives but instead find a way to keep it open so that different tooling implementations could add their own without further language changes.  For example, ignore any directive that starts with `##`, e.g. `##sdk Foo` or `##r "System.Windows.Forms` or `##load "script.cs"`.  Or just stop erroring on unrecognized directives outright and let the tooling warn/error if a particular directive is not recognized.

**Reply by jaredpar** [reply]
> The problem I see is that #sdk and #package doesn't really have a purpose except to be an implementation-defined comment (that most implementations don't really care about

Every compiler host that the .NET Team ships will definitely care about these directives. 

> The prior art here, even in the .NET ecosystem, usually goes for attributes (or in some cases pragmas or comments). 

The virtually no existing attribute or `#pragma` that controls or modifies _inputs_ to the compilation. That is exclusively the domain of the project file and build system. The one exception to this I'm aware of is strong naming. There are a few attributes which change the key used in strong naming but those are considered an anti-pattern in the compiler. Not something we would do again as negatively impacts our goals around input tracking and determinism.

> The attribute method is actually nicer in some respects since it potentially adds reflectable metadata, 

`<PackageReference>` exists today and serves the same purpose as `#package` yet there are no active requests for this to be reflected in metadata. 

> If we did want to keep it to shorter preprocessor directives I'd rather the language not adopt these specified directives but instead find a way to keep it open so that different tooling implementations could add their own without further language changes

The last time this came up, ~2-3 years ago, there was talk of just making `#!` the single _implementation defined_ directive. Compiler will parse `#!` and put the rest of the line content into effectively trivia. Tools like `dotnet run app.cs` could use `#! package` and go that route. I'm sure that will come up again when we discuss this in LDM.

**Reply by yaakov-h** [reply]
Has `#!/usr/bin/env dotnet run <options here>` been considered?

e.g. `#!/usr/bin/env dotnet run -f net10.0 --with-package System.Windows.Forms`

**Reply by jaredpar** [reply]
> Has #!/usr/bin/env dotnet run <options here> been considered?

Yes that is part of the feature set we're considering. It's not in our immediate plans, trying to nail some of the other core scenarios first, but it is on the backlog of items we want to look into.

**Reply by jkrejcha** [reply]
> The virtually no existing attribute or `#pragma` that controls or modifies inputs to the compilation.

Parts of the runtime arguably do (source generators), but this is besides the point. There's really nothing special about "it might include other files" that means that `#pragma` is insufficient. Regardless I'm not arguing for making whatever an attribute, as you say `<PackageReference>` is sufficient. But that's kinda part of the problem here, isn't it?

My argument is **against `#sdk` and `#package`** and **for a generalized solution** which could take the form of one or any of:
* Do nothing (and use `#pragma`)
* Add a `##` or `#!` or whatever or equivalent
* Allow conditionally supported directives

The specific directives don't add anything to the language except for another keyword for the compiler to ignore and for *maybe* the tooling to use.

I don't find the "this pragma doesn't exist today" argument convincing because well in this case the ball is in .NET's court here. If they want to do something unprecedented, they can. If they want to do nothing, they also can. It could exist tomorrow if it were simply willed into existence via the necessary code. That's what's useful about `#pragma`, implementations can do things specific to their ecosystem.

Like, the language doesn't (and shouldn't) have a concept of a package or SDK and there are good reasons for this. Not all implementations (that are in-use, if a bit more niche than .NET) support such a concept, nor should they be required to.

**Reply by jaredpar** [reply]
> Parts of the runtime arguably do (source generators), but this is besides the point

Can you list an example? How does the runtime implicitly add or modify inputs to the compilations? How are source generators relevant to this discussion? Source generators, ideally, follow the rules of the compiler with respect to inputs (don't read anything that wasn't explicitly passed to the compilation). Yes some violate this but that is considered a bug. 

>  If they want to do something unprecedented, they can

Sure we can ... but we have to ask if doing so fits into or subverts users expectations. Given that `#pragma` today is purely about controlling user diagnostics, having it suddenly be a way to change inputs to the compilation is very likely to fit into the subverts expectations category.

> Like, the language doesn't (and shouldn't) have a concept of a package or SDK and there are good reasons for this. 

This proposal doesn't change the C# language to understand packages. It is stating that it supports a directive that compilation tools can choose to, or choose not to, use. Similar to `#region` where the compiler hosts in IDEs leverage the directive but the command line compiler largely ignores it.

**Reply by 333fred** [reply]
> Do nothing (and use #pragma)

This option is simply a non-starter, as has been detailed numerous times in this discussion. Shells use `#!`. We don't control that. At the barest minimum, that directive must be added.

**Reply by jkrejcha** [reply]
Re source generators: they add new source that's then compiled (regex gen, libraryimport, etc). They're relevant if only to say that "adding a magic attribute adds a compilation unit" isn't unprecedented, in fact, it's pretty standard, and in fact I'd say the major current use of source generators. That's not against spec or anything, compilation units aren't limited to on-disk files. I feel like though there is some miscommunication here and it's more illustrative that similar things exist in .NET already and used the existing tooling just fine.

It's orthogonal to my main point ultimately.

---

> Sure we can ... but we have to ask if doing so fits into or subverts users expectations.

Which is that `#pragma` does something implementation-defined and not much more than that. That it happens to be useful for controlling diagnostics is a specific feature .NET provides along with... I guess checksumming. Especially considering that "pragma" generally is understood to mean "controls how input is processed in a implementation-defined way", I don't really see how this is all that confusing.

> This proposal doesn't change the C# language to understand packages. It is stating that it supports a directive that compilation tools can choose to, or choose not to, use. 

I think this is partially my point. There's no defined semantics for packages (and nor should there be). **Why then should the language have anything to say about it?**

---

To be clear, I'm actually pretty in favor of something like conditionally-supported directives. It'd change the error state of unknown directives to something more of an implementation-defined thing and increases backwards compatibility somewhat, and solves the fact that the *nix-like OS loader cares about shebangs for those actually running C# code as a script or whatever

It also has the advantage of not tying these directives to the language version and makes it so that if you want a snazzy new directive you don't have to go to LDM every single time. 

The only disadvantage ig is that pragma probably becomes much less attractive as an option even where it's the right choice (which would probably be most of the time).

**Reply by jkrejcha** [reply]
> This option is simply a non-starter, as has been detailed numerous times in this discussion. Shells use `#!`. We don't control that. At the barest minimum, that directive must be added.

It's the status quo... most programming languages don't special case this (iirc Python used to do some weird related hack at some point for DOS (the operating system) related reasons but they dropped that a very very long time ago). Non-starter seems a lil much in my view.

Imo, either supporting the prefix for implementation-defined anything or just supporting anything (in implementation-defined manner if unknown) imo would be much less hacky and piecemeal.

**Reply by jaredpar** [reply]
> Re source generators: they add new source that's then compiled (regex gen, libraryimport, etc) 

Yes they add source as a code object but this is 100% internal to the compilation process. There is no reaching out of the compilation and grabbing new artifacts off of disk. The source generator, in combination with the other compiler inputs, perfectly describes the inputs to the compilation. 

> Especially considering that "pragma" generally is understood to mean "controls how input is processed in a implementation-defined way"

This claim is simply incorrect. 

> There's no defined semantics for packages (and nor should there be). Why then should the language have anything to say about it?

The language has nothing to say about it, just as the language has nothing to say about `#region`. It is just a directive the language says that the host can define semantics for.

**Reply by jkrejcha** [reply]
> Yes they add source as a code object but this is 100% internal to the compilation process. There is no reaching out of the compilation and grabbing new artifacts off of disk. The source generator, in combination with the other compiler inputs, perfectly describes the inputs to the compilation.

There's no fundamental difference here.

> This claim is simply incorrect.

I'm kinda unsure what you're thinking here, given that most references to `pragma`, in general, say something to the effect of it being implementation defined. Given that .NET uses it outside of diagnostic control as well, I'd say that tracks with usage of the term both in C# as well as other programming languages and tools like C, C++, SQLite, etc.

> The language has nothing to say about it, just as the language has nothing to say about `#region`.

This is much more fundamentally different than region and even so, @HaloFour made [a pretty good point regarding this which bears repeating](https://github.com/dotnet/csharplang/discussions/8616#discussioncomment-12453287)

> You could make that argument, and honestly if someone had suggested `#region` much more recently it'd probably be a tougher sell. But at least it does have a specific meaning and use case, plus a known syntax (e.g. `#endregion` required).
> 
> The language can't know what these other directives mean, or if they have a specific structure, unless the language becomes much more bound to this specific tooling. It can't even know what directives are possible, or might be possible in a future release. Supporting a pattern for those directives and treating them like comments is probably the safest bet. But at the same time is there a reason that the tooling _must_ use new directives like this?

**Reply by jkrejcha** [reply]
(Regardless, the current proposal seems to have been changed. The current version (`#:` prefix) seems reasonable enough.)

**Reply by roke-julian-lockhart** [reply]
> [`#discussioncomment-12497852`](https://github.com/dotnet/csharplang/discussions/8616#discussioncomment-12497852)

@jkrejcha, could you link to that? I'm having trouble locating it.

**Reply by jjonescz** [reply]
I've just merged the update in https://github.com/dotnet/csharplang/pull/9215. So the new version is at https://github.com/dotnet/csharplang/blob/main/proposals/ignored-directives.md

**Reply by HaloFour** [reply]
Here's the PR where they reversed the order to have a known preprocessor prefix that the compiler always ignores as the primary proposal where known directives could be an alternative:

https://github.com/dotnet/csharplang/pull/9215

A few alternative sigils to `#:` have also been proposed.

**Comment by jkrejcha** [comment]
To be honest, I don't really seem the advantage this has *in the language* over just using `pragma` since this is already implementation specific. 

The idea of `sdk` and `package` is effectively specific only to the .NET implementation (the tooling)

**Reply by jkrejcha** [reply]
The only weird one is `#!` but even that could just be an extension used by CSX or something (it already has its own implementation with `#r`). 

~~Interestingly of note, it does seem that .NET effectively just ignores compiler directives it doesn't know about.~~ (See below)

**Reply by jjonescz** [reply]
> it does seem that .NET effectively just ignores compiler directives it doesn't know about.

If by "ignore" you mean "warns about", that's true but only for pragma directives, not arbitrary directives, i.e.:

```cs
#pragma test // warning CS1633: Unrecognized #pragma directive
#test // error CS1024: Preprocessor directive expected
```

**Reply by jkrejcha** [reply]
Ah ya, I checked again and it seemed I had a different file open when I tested the latter. Whoops.

**Comment by egvijayanand** [comment]
I would suggest adding another directive along with the `sdk`, `property`, and `package` to reference other C# source files as mentioned below.

The proposed `#:include` directive would add multiple C# source files to the compilation unit.

The file provided as a parameter to the `dotnet run` command would be the entry point, and other files are supporting files.

The source file path can be mentioned as an absolute path or a relative path to the current file.

```cs
#:include Foo.cs

// Other statements
```
